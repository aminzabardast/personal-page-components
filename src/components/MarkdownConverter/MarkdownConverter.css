.az-md-converter img {
    width: 100%;
}