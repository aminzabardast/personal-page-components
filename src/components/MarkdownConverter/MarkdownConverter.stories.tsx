import { StoryObj, Meta } from '@storybook/react'

import MarkdownConverter from './MarkdownConverter'

const meta: Meta<typeof MarkdownConverter> = {
    title: 'Blog/MarkdownConverter',
    component: MarkdownConverter,
}

export default meta

type Story = StoryObj<typeof meta>

export const Example1: Story = {
    args: {
        categories: [
            {
                title: 'Category Text 1',
                url: 'https://google.com/',
            },
            {
                title: 'Category Text 2',
                url: 'https://google.com/',
            },
            {
                title: 'Category Text 3',
                url: 'https://google.com/',
            },
            {
                title: 'Category Text 4',
                url: 'https://google.com/',
            },
        ],
        title: 'This is a Title',
        author: {
            name: 'Amin Zabardast',
            imageUrl:
                'https://raw.githubusercontent.com/aminzabardast/AminZabardast.com/master/src/assets/amin2.jpg',
        },
        date: new Date(),
        fileUrl:
            'https://raw.githubusercontent.com/highlightjs/highlight.js/main/README.md',
        mediumUrl: 'https://medium.com/',
    },
}

export const Example2: Story = {
    args: {
        categories: [
            {
                title: 'Category Text 1',
                url: 'https://google.com/',
            },
            {
                title: 'Category Text 2',
                url: 'https://google.com/',
            },
            {
                title: 'Category Text 3',
                url: 'https://google.com/',
            },
            {
                title: 'Category Text 4',
                url: 'https://google.com/',
            },
        ],
        title: 'This is a Title',
        author: {
            name: 'Amin Zabardast',
            imageUrl:
                'https://raw.githubusercontent.com/aminzabardast/AminZabardast.com/master/src/assets/amin2.jpg',
        },
        date: new Date(),
        fileUrl:
            'https://raw.githubusercontent.com/aminzabardast/AminZabardast.com/gh-pages/blog_posts/2021/04/24/git-gamification.md',
        mediumUrl: 'https://medium.com/',
    },
}
