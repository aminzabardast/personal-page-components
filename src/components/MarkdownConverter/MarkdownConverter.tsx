import React, { useState } from 'react'
import { Converter } from 'showdown'
import CircularProgress from '@mui/material/CircularProgress'
import Box from '@mui/material/Box'
import Container from '@mui/material/Container'
import Typography from '@mui/material/Typography'
import Stack from '@mui/material/Stack'
import Divider from '@mui/material/Divider'
import Avatar from '@mui/material/Avatar'
import Chip from '@mui/material/Chip'
import { format } from 'date-fns'
import './MarkdownConverter.css'
import { mdiCommentOutline } from '@mdi/js'
import Button from '@mui/material/Button'
import SvgIcon from '@mui/material/SvgIcon'

type Category = {
    title: string
    url: string
}

type Author = {
    name: string
    imageUrl?: string
}

interface MarkdownConverterProps {
    categories?: Category[]
    title: string
    date?: Date
    author?: Author
    fileUrl: string
    mediumUrl?: string
}

const MarkdownConverter = ({
    categories,
    title,
    date,
    author,
    fileUrl,
    mediumUrl,
}: MarkdownConverterProps) => {
    const [isLoading, setIsLoading] = useState(true)

    const [markdownText, setMarkdownText] = useState('')
    fetch(fileUrl).then((response) => {
        response.text().then((text) => {
            setMarkdownText(text)
            setIsLoading(false)
        })
    })

    const converter = new Converter({
        extensions: [],
    })

    return (
        <div data-testid="az-blog-object">
            {title && (
                <Container component="div">
                    <Typography variant="h1">{title}</Typography>
                </Container>
            )}
            {(date || author) && (
                <Container>
                    <Stack
                        direction="row"
                        divider={<Divider orientation="vertical" flexItem />}
                        spacing={2}
                    >
                        {author && (
                            <Stack direction="row" spacing={2}>
                                {author?.imageUrl && (
                                    <Avatar
                                        src={author.imageUrl}
                                        sx={{ width: 32, height: 32 }}
                                    />
                                )}
                                {author?.name && (
                                    <Typography variant="h5" component="span">
                                        {author.name}
                                    </Typography>
                                )}
                            </Stack>
                        )}
                        {date && (
                            <Typography variant="h5" component="span">
                                {format(date, 'PPP')}
                            </Typography>
                        )}
                    </Stack>
                </Container>
            )}
            <Container sx={{ paddingTop: '40px' }}>
                <Divider />
            </Container>
            {isLoading && (
                <Box display="flex" justifyContent="center" alignItems="center">
                    <CircularProgress />
                </Box>
            )}
            {!isLoading && (
                <Container
                    className="az-md-converter"
                    component="div"
                    dangerouslySetInnerHTML={{
                        __html: converter.makeHtml(markdownText),
                    }}
                ></Container>
            )}
            {categories && (
                <Container sx={{ paddingTop: '40px' }}>
                    <Stack direction={{ xs: 'column', sm: 'row' }} spacing={2}>
                        {categories.map((category) => {
                            return (
                                <Chip
                                    data-testid="az-categories"
                                    key={category.title}
                                    label={category.title}
                                    component="a"
                                    href={category.url}
                                    clickable
                                />
                            )
                        })}
                    </Stack>
                </Container>
            )}
            {mediumUrl && (
                <Container sx={{ paddingTop: '40px' }}>
                    <Button
                        component="a"
                        href={mediumUrl}
                        data-testid="az-medium-link"
                        target="_blank"
                        style={{
                            textTransform: 'none',
                        }}
                        startIcon={
                            <SvgIcon>
                                <path d={mdiCommentOutline}></path>
                            </SvgIcon>
                        }
                    >
                        Leave a comment on Medium
                    </Button>
                </Container>
            )}
        </div>
    )
}

export default MarkdownConverter
