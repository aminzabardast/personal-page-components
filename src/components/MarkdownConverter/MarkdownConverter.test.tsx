import React from 'react'
import { render, waitFor, screen } from '@testing-library/react'
import MarkdownConverter from './MarkdownConverter'
import '@testing-library/jest-dom'
import renderer from 'react-test-renderer'

global.fetch = jest.fn(() =>
    Promise.resolve({
        text: () => Promise.resolve('# Header1'),
    })
) as jest.Mock

const examplePost = (
    <MarkdownConverter
        title="This is a Title"
        categories={[
            {
                title: 'Category1',
                url: 'https://aminzabardast.com',
            },
            {
                title: 'Category2',
                url: 'https://aminzabardast.com',
            },
            {
                title: 'Category3',
                url: 'https://aminzabardast.com',
            },
        ]}
        date={new Date('2000-01-01T00:00:00')}
        author={{
            name: 'Amin Zabardast',
            imageUrl: 'https://aminzabardast.com/image.png',
        }}
        fileUrl="https://blog.aminzabardast.com/someMarkDownFile.md"
    />
)

describe('<MarkdownConverter />', () => {
    it('should render.', async () => {
        render(examplePost)
        await waitFor(() => {
            expect(screen.getByTestId('az-blog-object')).toBeInTheDocument()
        })
    })
    it('should show the title.', async () => {
        render(examplePost)
        await waitFor(() => {
            expect(screen.getByText('This is a Title')).toBeInTheDocument()
        })
    })
    it('should show the category titles.', async () => {
        render(examplePost)
        await waitFor(() => {
            expect(screen.getAllByTestId('az-categories').length).toEqual(3)
        })
    })
    it('should show the category as a link.', async () => {
        render(examplePost)
        await waitFor(() => {
            screen
                .getAllByTestId('az-categories')
                .forEach((item) =>
                    expect(item).toHaveAttribute(
                        'href',
                        'https://aminzabardast.com'
                    )
                )
        })
    })
    it('should show the date.', async () => {
        render(examplePost)
        await waitFor(() => {
            expect(screen.getByText('January 1st, 2000')).toBeInTheDocument()
        })
    })
    it('should show the authors name.', async () => {
        render(examplePost)
        await waitFor(() => {
            expect(screen.getByText('Amin Zabardast')).toBeInTheDocument()
        })
    })
    it('should have the authors image.', async () => {
        render(examplePost)
        await waitFor(() => {
            expect(screen.getByRole('img')).toHaveAttribute(
                'src',
                'https://aminzabardast.com/image.png'
            )
        })
    })
    it('should get the markdown file and render its content.', async () => {
        render(examplePost)
        await waitFor(() => {
            expect(screen.getByText('Header1')).toBeInTheDocument()
        })
    })
})
describe('<MarkdownConverter /> SnapShot Test', () => {
    it('should keep producing consistent result.', async () => {
        const tree = renderer.create(examplePost).toJSON()
        await waitFor(() => {
            expect(tree).toMatchSnapshot()
        })
    })
})
