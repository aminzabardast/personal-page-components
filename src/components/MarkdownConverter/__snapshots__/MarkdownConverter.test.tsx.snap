// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MarkdownConverter /> SnapShot Test should keep producing consistent result. 1`] = `
<div
  data-testid="az-blog-object"
>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  >
    <h1
      className="MuiTypography-root MuiTypography-h1 css-1q74aok-MuiTypography-root"
    >
      This is a Title
    </h1>
  </div>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  >
    <div
      className="MuiStack-root css-pl8gqh-MuiStack-root"
    >
      <div
        className="MuiStack-root css-pl8gqh-MuiStack-root"
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular css-xwmgj5-MuiAvatar-root"
        >
          <img
            className="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="https://aminzabardast.com/image.png"
          />
        </div>
        <span
          className="MuiTypography-root MuiTypography-h5 css-1wn2x41-MuiTypography-root"
        >
          Amin Zabardast
        </span>
      </div>
      <hr
        className="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
      />
      <span
        className="MuiTypography-root MuiTypography-h5 css-1wn2x41-MuiTypography-root"
      >
        January 1st, 2000
      </span>
    </div>
  </div>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg css-76a484-MuiContainer-root"
  >
    <hr
      className="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
  </div>
  <div
    className="MuiBox-root css-1n2mv2k"
  >
    <span
      className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
      role="progressbar"
      style={
        {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={{}}
        />
      </svg>
    </span>
  </div>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg css-76a484-MuiContainer-root"
  >
    <div
      className="MuiStack-root css-w7gobd-MuiStack-root"
    >
      <a
        className="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-1b2taxe-MuiButtonBase-root-MuiChip-root"
        data-testid="az-categories"
        href="https://aminzabardast.com"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
      >
        <span
          className="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
        >
          Category1
        </span>
      </a>
      <a
        className="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-1b2taxe-MuiButtonBase-root-MuiChip-root"
        data-testid="az-categories"
        href="https://aminzabardast.com"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
      >
        <span
          className="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
        >
          Category2
        </span>
      </a>
      <a
        className="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-filledDefault css-1b2taxe-MuiButtonBase-root-MuiChip-root"
        data-testid="az-categories"
        href="https://aminzabardast.com"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
      >
        <span
          className="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
        >
          Category3
        </span>
      </a>
    </div>
  </div>
</div>
`;
